node/overseer/src/lib.rs

```diff
@@ -399,13 +424,17 @@ pub struct Overseer<S: SpawnNamed> {
 	/// Events that are sent to the overseer from the outside world
 	events_rx: mpsc::Receiver<Event>,
 
+	/// External listeners waiting for a hash to be in the active-leave set.
+	// TODO (now): how to clean it up? Use LRUCache? Or .retain(|_, v| { v.retain(|c| !c.is_cancelled()); v })
+	activation_external_listeners: HashMap<Hash, Vec<oneshot::Sender<()>>>,
```

Comment:
`LRUCache` can't be the right answer because in the event that lots of hashes are interesting, some will drop, breaking the provisioner for those hashes. That said, you're right: we need some kind of long-term cleanup here, because otherwise we will see unbounded growth in two cases:

- someone registers a listener for a hash which used to be in the active set, but no longer is
- someone registers a listener for an invalid / broken hash

I don't know what the best solution is. My best idea so far is to leverage finalization. It goes like this:

- edit the definition here:
   ```rust
   activation_external_listeners: HashMap<Hash, (BlockNumber, Vec<oneshot::Sender<()>>)>,
   ```
- when adding to the listener list (line 1154 as of this commit), figure out the block number (runtime api request, probably) associated with the supplied hash. If no block number is available, don't actually add the listener to the list.
- `on_finalized`: `activation_external_listeners.retain(|_, (listener_block_number, _)| listener_block_number >= finalized.block_number);`

Reply:
The idea of leveraging finalization sounds interesting, but I bet we don't want to make runtime api requests here and in practice it could be the case that finalized block is far behind.

Since the proposer has its own timeout, we could simply periodically call `.retain(|_, v| { v.retain(|c| !c.is_cancelled()); !v.is_empty() })` e.g. on `block_import` and that would clean (assuming the receiver was dropped) or utilize something like https://github.com/stusmall/ttl_cache instead of a hashmap.